File: c/Binary-Search.c
#include "binary_search.h"
const int *binary_search(int value, const int *arr, size_t length) {
    int low = 0;
    // 'length' size_t türünde olduğundan, aritmetik işlemler için int'e dönüştürülmesi iyi bir uygulamadır
    int high = (int)length - 1;

    while (low <= high) {
        int mid = low + (high - low) / 2; // Orta indeksi hesapla

        // Değer orta indekste bulunduysa
        if (arr[mid] == value) {
            return &arr[mid]; // Bulunan değerin adresini döndür
        }
        // Aranacak değer ortadaki değerden büyükse, sağ yarıyı ara
        else if (arr[mid] < value) {
            low = mid + 1;
        }
        // Aranacak değer ortadaki değerden küçükse, sol yarıyı ara
        else {
            high = mid - 1;
        }
    }
    return NULL; // Değer bulunamadıysa NULL döndür
}

File: python/BinarySearch.py
def find(search_list, value):
  """
  Bir sıralı listede ikili arama (binary search) yaparak bir değeri bulur.

  Args:
    search_list: Aranacak sıralı tam sayı listesi.
    value: Aranacak tam sayı değeri.

  Returns:
    Değer bulunursa, o değerin listedeki indeksini döndürür.

  Raises:
    ValueError: Değer listede bulunamazsa "value not in array" mesajıyla bir hata fırlatır.
  """
  low = 0
  high = len(search_list) - 1

  while low <= high:
    # Orta indeksi hesapla
    # high + low yerine low + (high - low) // 2 kullanmak,
    # çok büyük sayılarda olası taşma (overflow) sorununu önler,
    # ancak Python'da tam sayı boyutu sınırı olmadığı için burada kritik değil.
    mid = low + (high - low) // 2

    # Eğer aranan değer ortadaki elemandan büyükse, sol yarıyı yok say.
    if search_list[mid] < value:
      low = mid + 1

    # Eğer aranan değer ortadaki elemandan küçükse, sağ yarıyı yok say.
    elif search_list[mid] > value:
      high = mid - 1

    # Eğer aranan değer ortadaki elemana eşitse, değer bulundu!
    else:
      return mid

  # Döngü biterse ve buraya ulaşılırsa, bu demektir ki eleman listede yoktu.
  # Bu durumda belirtilen ValueError'ı fırlatmalıyız.
  raise ValueError("value not in array")